File: Lecture Demos/Week 6/3. Matches Demo/Python/finding_matches.py
#A more advanced method using list comprehension
#return [result for item in collection if condition]
def zWords2(words):
    return [word.upper() for word in words if 'z' in word]

def is_z_word(word):
    return 'z' in word

def zWords(words):
    matches = []
    for word in words:
        if is_z_word(word):
            matches.append(word)
    return matches

words = ["snooze", "david", "zed", "dyer"]
print(f"Matching words: \n{zWords(words)}")

File: Lecture Demos/Week 6/3. Matches Demo/Java/FindingMatches.java
import java.util.Arrays;
import java.util.Iterator;
import java.util.LinkedList;

public class FindingMatches {
    public static void main(String[] args) {
        LinkedList<String> words = new LinkedList<String>(Arrays.asList("snooze", "david", "zed", "dyer"));
        
        System.out.println("List before: ");
        for (String word : words) { System.out.println(word); }
        
        removeZWords(words);
        
        System.out.println("List after: ");
        for (String word : words) { System.out.println(word); }
    }

    //Useful method for returning the matches themself
    // private static LinkedList<String> zWords(LinkedList<String> words) {
    //     LinkedList<String> matches = new LinkedList<String>();
    //     for (String word : words) {
    //         if (word.contains("z")) {
    //             matches.add(word);
    //         }
    //     }
    //     return matches;
    // }

    private static boolean isZWord(String word) {
        return word.contains("z");
    }

    private static void removeZWords(LinkedList<String> words) {
        LinkedList<String> wordsToRemove = new LinkedList<String>();
        for (String word : words) {
            if (isZWord(word)) {
                wordsToRemove.add(word);
            }
        }
        words.removeAll(wordsToRemove);
    }
}